src/main/java/org/generationcp/middleware/api/germplasm/GermplasmServiceImpl.java

```diff
@@ -0,0 +1,166 @@
+
+			//First iteration, parents not provided, default = 0
+			germplasm.setGpid1(0);
+			germplasm.setGpid2(0);
+
+			germplasm.setGrplce(0);
+			germplasm.setMgid(0);
+			germplasm.setUserId(userId);
+			germplasm.setLgid(0);
+			germplasm.setLocationId(locationsMapByAbbr.get(germplasmDto.getLocationAbbr().toUpperCase()));
+			germplasm.setDeleted(Boolean.FALSE);
+			germplasm.setGdate(Util.convertDateToIntegerValue(germplasmDto.getCreationDate()));
+
+			if (StringUtils.isEmpty(germplasmDto.getGuid())) {
+				GermplasmGuidGenerator.generateGermplasmGuids(cropType, Collections.singletonList(germplasm));
+			} else {
+				germplasm.setGermplasmUUID(germplasmDto.getGuid());
+			}
+
+			if (!StringUtils.isEmpty(germplasmDto.getReference())) {
+				final Bibref bibref =
+					new Bibref(null, DEFAULT_DASH, DEFAULT_DASH, germplasmDto.getReference(), DEFAULT_DASH, DEFAULT_DASH, DEFAULT_DASH,
+						DEFAULT_DASH,
+						DEFAULT_DASH, DEFAULT_DASH, DEFAULT_DASH, DEFAULT_DASH);
+				this.daoFactory.getBibrefDAO().save(bibref);
+				germplasm.setReferenceId(bibref.getRefid());
+			} else {
+				germplasm.setReferenceId(0);
+			}
+
+			this.daoFactory.getGermplasmDao().save(germplasm);
+
+			germplasmDto.getNames().forEach((k, v) -> {
+				final Name name = new Name(null, germplasm.getGid(), nameTypesMapByName.get(k.toUpperCase()),
+					(k.equalsIgnoreCase(germplasmDto.getPreferredName())) ? 1 : 0, userId, v, germplasm.getLocationId(),
+					Util.getCurrentDateAsIntegerValue(), 0);
+				this.daoFactory.getNameDao().save(name);
+			});
+
+			if (germplasmDto.getAttributes() != null) {
+				germplasmDto.getAttributes().forEach((k, v) -> {
+					final Attribute attribute = new Attribute(null, germplasm.getGid(), attributesMapByName.get(k.toUpperCase()), userId, v,
+						germplasm.getLocationId(),
+						0, Util.getCurrentDateAsIntegerValue());
+					this.daoFactory.getAttributeDAO().save(attribute);
+				});
+			}
+			results.put(germplasmDto.getClientId(),
+				new GermplasmImportResponseDto(GermplasmImportResponseDto.Status.CREATED, Collections.singletonList(germplasm.getGid())));
+		}
+
+		return results;
+	}
+
+	private Map<String, Integer> getLocationsMapByAbbr(final List<GermplasmImportRequestDto.GermplasmDto> germplasmDtos) {
+		final Set<String> locationAbbreviations = germplasmDtos.stream().map(g -> g.getLocationAbbr()).collect(Collectors.toSet());
+		return this.daoFactory.getLocationDAO().getByAbbreviations(new ArrayList<>(locationAbbreviations)).stream()
+			.collect(Collectors.toMap(Location::getLabbr, Location::getLocid));
+	}
+
+	private Map<String, Method> getBreedingMethodsMapByAbbr(final List<GermplasmImportRequestDto.GermplasmDto> germplasmDtos) {
+		final Set<String> breedingMethods = germplasmDtos.stream().map(g -> g.getBreedingMethodAbbr()).collect(Collectors.toSet());
+		return this.daoFactory.getMethodDAO().getByCode(new ArrayList<>(breedingMethods)).stream()
+			.collect(Collectors.toMap(Method::getMcode, method -> method));
+	}
```

Comment:
Some of these methods look useful, maybe we can have them in the daos

Reply:
Agree that may be useful but they depend on the user input, so I would not located in the DAO's. I saw in Aldrin's branch similar code, so when bot are merged we can find a way to reuse them